src/types/set.rs

```diff
@@ -71,96 +71,248 @@ impl PySet {
     where
         K: ToPyObject,
     {
-        fn inner(set: &PySet, key: PyObject) -> PyResult<bool> {
+        Bound::borrowed_from_gil_ref(&self).contains(key)
+    }
+
+    /// Removes the element from the set if it is present.
+    ///
+    /// Returns `true` if the element was present in the set.
+    pub fn discard<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToPyObject,
+    {
+        Bound::borrowed_from_gil_ref(&self).discard(key)
+    }
+
+    /// Adds an element to the set.
+    pub fn add<K>(&self, key: K) -> PyResult<()>
+    where
+        K: ToPyObject,
+    {
+        Bound::borrowed_from_gil_ref(&self).add(key)
+    }
+
+    /// Removes and returns an arbitrary element from the set.
+    pub fn pop(&self) -> Option<PyObject> {
+        Bound::borrowed_from_gil_ref(&self).pop().map(|e| e.into())
+    }
+
+    /// Returns an iterator of values in this set.
+    ///
+    /// # Panics
+    ///
+    /// If PyO3 detects that the set is mutated during iteration, it will panic.
+    pub fn iter(&self) -> PySetIterator<'_> {
+        IntoIterator::into_iter(self)
+    }
+}
+
+/// Implementation of functionality for [`PySet`].
+///
+/// These methods are defined for the `Bound<'py, PySet>` smart pointer, so to use method call
+/// syntax these methods are separated into a trait, because stable Rust does not yet support
+/// `arbitrary_self_types`.
+#[doc(alias = "PySet")]
+pub trait PySetMethods<'py> {
+    /// Removes all elements from the set.
+    fn clear(&self);
+
+    /// Returns the number of items in the set.
+    ///
+    /// This is equivalent to the Python expression `len(self)`.
+    fn len(&self) -> usize;
+
+    /// Checks if set is empty.
+    fn is_empty(&self) -> bool;
+
+    /// Determines if the set contains the specified key.
+    ///
+    /// This is equivalent to the Python expression `key in self`.
+    fn contains<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToPyObject;
+
+    /// Removes the element from the set if it is present.
+    ///
+    /// Returns `true` if the element was present in the set.
+    fn discard<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToPyObject;
+
+    /// Adds an element to the set.
+    fn add<K>(&self, key: K) -> PyResult<()>
+    where
+        K: ToPyObject;
+
+    /// Removes and returns an arbitrary element from the set.
+    fn pop(&self) -> Option<Bound<'py, PyAny>>;
+
+    /// Returns an iterator of values in this set.
+    ///
+    /// # Panics
+    ///
+    /// If PyO3 detects that the set is mutated during iteration, it will panic.
+    fn iter(&self) -> BoundSetIterator<'py>;
+}
+
+impl<'py> PySetMethods<'py> for Bound<'py, PySet> {
+    #[inline]
+    fn clear(&self) {
+        unsafe {
+            ffi::PySet_Clear(self.as_ptr());
+        }
+    }
+
+    #[inline]
+    fn len(&self) -> usize {
+        unsafe { ffi::PySet_Size(self.as_ptr()) as usize }
+    }
+
+    fn is_empty(&self) -> bool {
```

Comment:
Can you clarify what you mean, I assume you want me to move this method body into the trait definition?

(I'm not sure there's any material value in doing so; I felt that it made sense to keep all the method bodies together.)

Reply:
Yes, I meant moving the body into the trait.

I think this would make sense because of two reasons:
* The implementation will be the same for all implementors (i.e. if we have future variants of `impl PyFooMethods`. I can't find, but I think we already hit a situation where additional impls were discussed.), so it can be shared.
* The implementation is of a "different kind" than the other ones. It does not really call into the FFI but rather it is a convenience method defined purely via other methods in the trait. So I think it would actually make sense to separate them to avoid adding noise to the block containing the "real" implementations.

Reply:
Done 👍